### ISSUE TITLE
Rule Proposal: `prefer-smaller-scope`

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#5284]
Currently `prefer-const` has two behaviors.
1. detecting it's never modified.
2. detecting it can be moved to a smaller scope.

This seems to be confusing people (e.g. #5283, #5251) (I'm sorry...).
So this proposal is separating the 2nd behavior to new rule: `prefer-smaller-scope`.

`prefer-small-scope` is aimed at eliminating variables in too wide scope.
This rule warns variables which can move to a smaller scope.
This improves readability because the scope which can read/write the variable gets smaller.

The following patterns are considered problems:

``` js
/* eslint prefer-smaller-scope: 2 */

let foo;        /*error 'foo' declaration can move into a block 'if (a) {...}'.*/
if (a) {
    foo = 100;
    bar(foo);
} else {
    bar(200);
}
```

The following patterns are not considered problems:

``` js
/* eslint prefer-smaller-scope: 2 */

let foo;
if (a) {
    foo = 100;
} else {
    foo = 200:
}
baz(foo);

let foo;
try {
    foo = bar();
} catch (err) {
    foo = 0:
}
baz(foo);
```

Then the 2nd behavior of `prefer-const` is removed.
I think this may be not a breaking change because the rule is defused than now.

If accepted, I'm happy to work.

**Comment by nzakas** [comment]
I don't think this is the correct solution to this problem. In both of the referenced issues, the problem is that a `let` binding is created without initialization and is assigned later. Why don't we just stop warning on that case?

Example:

```js
let x;  // no init, don't warn
if (foo){
    x = 1;
}
```

**Comment by BYK** [comment]
@nzakas - Agree with your solution but I still think a rule like this would be useful.

**Comment by michaelficarra** [comment]
:+1: I like this proposal.

**Comment by nzakas** [comment]
Just so that I'm clear, the proposal here is also just to target `let` declarations that are uninitialized?

**Comment by michaelficarra** [comment]
@nzakas No, not at all.

> `prefer-small-scope` is aimed at eliminating variables in too wide scope. This rule warns variables which can move to a smaller scope. This improves readability because the scope which can read/write the variable gets smaller.

**Comment by nzakas** [comment]
@michaelficarra I read the original description, but the examples cited do not show any other instances. If you want to flesh out the proposal, then please show some other instances.

**Comment by michaelficarra** [comment]
@nzakas There's no reason class and function declarations wouldn't also be affected.

**Comment by mysticatea** [comment]
@nzakas Yes, I will stop warning in that case. Plus, I don't want to delete the 2nd behavior merely, so I made this proposal.

I'm thinking this rule applies to both `let` and `const`.
I will use a logic based on `prefer-const`, but I need to consider difference between real constants and readonly values.
I'm thinking: if all references in the scope are non-writable, it checks the moving doesn't go over the border of a loop or a function. If goes over that, it skips the variable.

For example, in the following pattern, the `fs` variable can move into the `read` function, but it goes over the border of a function, so it's not warned.

```js
const fs = require("fs");

function read(path) {
    return fs.readFileSync(path);
}
```

The following pattern is similar.

```js
const pattern = /abc(.+)/;

for (var i = 0; i < items.length; ++i) {
    if (pattern.test(items[i])) {
        foo(items[i]);
    }
}
```

The following pattern is not.

```js
const fs = require("fs");  /*error 'fs' declaration can move into a block 'if (a) {...}'.*/

let content;
// This block is not a function or a loop.
if (a) {
    content = fs.readFileSync(b);
} else {
    content = "";
}
```

```js
let item;  /*error 'item' declaration can move into a block 'while (items.length) {...}'.*/
// A writable reference exists in this block.
while (items.length) {
    item = items.pop();
}
```

**Comment by doberkofler** [comment]
I would really like this new rule

**Comment by nzakas** [comment]
I think `require` is a dangerous case for this rule, since a lot of people use all `require` statements at the top of the file.

**Comment by mysticatea** [comment]
> I'm thinking: if all references in the scope are non-writable, it checks the moving doesn't go over the border of a loop or a function. If goes over that, it skips the variable.

I think this way can handle `require()`s safely.

However, I can make an exception for `require()`s.

- If the variable has `require()` in their initializer, this rule skips the variable.

I hope to turn this exception off by an option (e.g. `{exceptRequires: false}`) when I use this rule (It can prevent importing actually unused modules, though, I know core doesn't have rules for performance...).

**Comment by nzakas** [comment]
OK, sounds good.

**Comment by indexzero** [comment]
@nzakas I was bit by this today, was a serious head scratcher for a while. Glad to hear it's getting attention from you fine folks. 

Two things:

1. I'm a big +1 on removing warnings from `prefer-const` for uninitialized `let` statements, is that still happening? It was a little unclear reading the back scroll. 
2. If you're looking for a common pattern where this happens [I see it a lot in `mocha` tests](https://gist.github.com/indexzero/0df4264d5a1121469d2e#file-prefer-const-fail-test-js) where there was some test fixture that is still being setup, but no longer consumed in tests. `prefer-smaller-scope: 1` would be a helpful way to point out they can be safely removed.

**Comment by nzakas** [comment]
> I'm a big +1 on removing warnings from prefer-const for uninitialized let statements, is that still happening?

Yes

**Comment by mysticatea** [comment]
I'm sorry for late.
I will send a PR within a few days.

**Comment by kaicataldo** [comment]
@mysticatea Looks like the `prefer-const` portion of the issue was merged - is the new `prefer-smaller-scope` rule still something we want?

**Comment by kaicataldo** [comment]
@mysticatea Where does this stand?

**Comment by mysticatea** [comment]
I'm sorry, I'm closing for now. 

I have encountered a hard problem. `prefer-smaller-scope` needs to move variable declarations, it's very difficult to distinguish if it can move the declarations because of references in their initializers. I will revisit this after I investigate the way of type/value tracking in our scope analysis.
